### PR TITLE
Add formal charge to the `to_dataframe()` and `from_dataframe()` methods

### PR DESCRIPTION
**Pull request by jchen0506**
This PR is to addressing issue #2001 
1. Atoms with formal charges are now supported by `to_dataframe()` and `from_dataframe()`
2. Unit tests for these two functions are provided

### CODE REVIEW

**Review by mattwthompson — APPROVED** [review]

File: mdtraj/core/topology.py [hunk 1]
```diff
@@ -641,13 +645,20 @@ def from_dataframe(
                     atom["resSeq"],
                     atom["segmentID"],
                 )
-
+            fc = atom["formal_charge"]
+            if fc is not None:
+                if pd.notna(fc):
+                    fc = float(fc)
```
Comment (janash) [inline mdtraj/core/topology.py:651]: Formal charges are integers by the PDB definition. I understand not wanting to cast as `int` in case someone has input a floating point number, but if something like `1` is cast as a float here it will cause problems down the line if you try to write a PDB (ie it will fail because the line is too long).

I think there should be a check here to make sure the charges can be cast as `int`, and then they should be cast as `int` rather than float.

File: mdtraj/core/topology.py [hunk 2]
```diff
@@ -592,6 +593,9 @@ def from_dataframe(
             if col not in atoms.columns:
                 raise ValueError("dataframe must have column %s" % col)
 
+        if "formal_charge" not in atoms.columns:
```
Comment (janash) [inline mdtraj/core/topology.py:596]: I'm not a huge fan of how this function modifies the user's dataframe, but there is already a precedent for this with `segmentID`

**Review by janash — CHANGES_REQUESTED** [review]
Sorry! I have a request on how we are casting the charge. It should be cast as an `int` rather than a float. 

It appears to be mislabeled [here](https://github.com/janash/mdtraj/blob/b7c3406f185a3496efbd1935189008bc826342b7/mdtraj/core/topology.py#L1662) (though it is labeled as `int` for the `add_atom` method). That could be fixed in this PR as well.

File: mdtraj/core/topology.py (hunk 1 above)
Comment (jchen0506) [reply mdtraj/core/topology.py]: Yes, I agree casting to int makes more sense. I've fixed the incorrect type annotation for the formal charge in atom and also added a check to make sure the charge can be  cast as `int`

File: mdtraj/core/topology.py [hunk 3]
```diff
@@ -531,6 +532,7 @@ def to_dataframe(self) -> tuple[pd.DataFrame, NDArray[np.float64] | None]:
                 "resName",
                 "chainID",
                 "segmentID",
+                "formal_charge"
```
Comment (janash) [inline mdtraj/core/topology.py:535]: I've been thinking a bit more about this. It seems like we'd want everything in this dataframe to be an integer or object data type (integer for index, object for name). I checked what is coming out now, and `formal_charge` is being upcast to a floating point number. Apparently this happens when you have `NaN` or `None` for some values in a column, but we could ensure that these are returned as the proper type by doing the following *after* dataframe creation (so my comment here isn't quite in the right spot. Using `Int64` with a capital `I` will allow Nan values - it is the [nullable integer type](https://pandas.pydata.org/docs/user_guide/integer_na.html) from pandas.

```
# Using Int64 makes the data type integer
# but allows for NA values in the column
atoms = atoms.astype({
    "formal_charge": "Int64"
})
```

File: mdtraj/core/topology.py (hunk 2 above)
Comment (janash) [reply mdtraj/core/topology.py:596]: It was modifying the dataframe before, so I think it's okay to *not* make a copy. It would change the behavior of the function, so let's leave it for now.

File: mdtraj/core/topology.py [hunk 4]
```diff
@@ -641,13 +645,23 @@ def from_dataframe(
                     atom["resSeq"],
                     atom["segmentID"],
                 )
-
+            fc = atom["formal_charge"]
+            if fc is not None:
```
Comment (janash) [inline mdtraj/core/topology.py:649]: Instead of having these if statements, you could cast to `Int64` before iterating over the atoms (with a try/except block and `astype` I mentioned above). I'd advise raising a `ValueError` in the `except`. I believe when casting to `Int64` it will ignore NaNs and also raise an error if a value that is not an integer is encountered. It will allow this to be done all at once (probably must faster than checking individual values) and will also be cleaner in the code.

**Review by janash — CHANGES_REQUESTED** [review]

File: mdtraj/core/topology.py (hunk 3 above)
Comment (jchen0506) [reply mdtraj/core/topology.py:535]: Thank you! This is much more cleaner. I've changed the code accordingly and add a normalize function in unit test to normalize the formal charge since after casting with `astype("formal_charge": "Int64")`, all None becomes <NA>

File: mdtraj/core/topology.py [hunk 5]
```diff
@@ -592,9 +597,19 @@ def from_dataframe(
...
+
+
         if "segmentID" not in atoms.columns:
             atoms["segmentID"] = ""
 
+        # Ensure formal_charge is of type Int64 (nullable integer)
+        try:
+            atoms = atoms.astype({"formal_charge": "Int64"})
+        except ValueError as e:
+            raise ValueError(f"Error converting formal_charge to Int64: {e}")
```
Comment (janash) [inline mdtraj/core/topology.py:611]: Could this say something else about checking to make sure your charges are integer values?

Comment (jchen0506) [reply mdtraj/core/topology.py]: Sure, thanks for spotting that

**Review by janash — APPROVED** [review]
Excellent. Thanks Jing! And thank you for your patience in the review.
